src/worker/mod.rs

```diff
@@ -176,18 +181,28 @@ impl<S: AsyncWrite + Send + Unpin, E> Worker<S, E> {
     }
 }
 
-enum Failed<E: StdError> {
+enum Failed<E: StdError, JE: StdError> {
     Application(E),
+    HandlerPanic(JE),
     BadJobType(String),
 }
 
 impl<S: AsyncBufRead + AsyncWrite + Send + Unpin, E: StdError + 'static + Send> Worker<S, E> {
-    async fn run_job(&mut self, job: Job) -> Result<(), Failed<E>> {
+    async fn run_job(&mut self, job: Job) -> Result<(), Failed<E, JoinError>> {
         let handler = self
             .callbacks
             .get(job.kind())
             .ok_or(Failed::BadJobType(job.kind().to_string()))?;
-        handler.run(job).await.map_err(Failed::Application)
+        match handler {
+            Callback::Async(cb) => cb.run(job).await.map_err(Failed::Application),
```

Comment:
I am worried that this is not the case for async handlers, panicking in there is not intercepted in the way is now being done for `register_blocking_fn`. I wonder if might need to use `futures::future::FutureExt::catch_unwind` for those futures, not sure..

Reply:
Let's do that in a follow-up (maybe the tech debt one?)

Reply:
pinned in there